src/hooks/useViewportRows.ts

```diff
@@ -0,0 +1,102 @@
+import { useMemo } from 'react';
+
+import { GroupRow, GroupByDictionary, Dictionary } from '../types';
+import { getVerticalRangeToRender } from '../utils';
+
+interface ViewportRowsArgs<R, SR> {
+  rawRows: readonly R[];
+  rowHeight: number;
+  clientHeight: number;
+  scrollTop: number;
+  groupBy: readonly string[];
+  rowGrouper?: (rows: readonly R[], columnKey: string) => Dictionary<readonly R[]>;
+  expandedGroupIds?: ReadonlySet<unknown>;
+}
+
+export function useViewportRows<R, SR>({
+  rawRows,
+  rowHeight,
+  clientHeight,
+  scrollTop,
+  groupBy,
+  rowGrouper,
+  expandedGroupIds
+}: ViewportRowsArgs<R, SR>) {
+  const [groupedRows, rowsCount] = useMemo(() => {
+    if (groupBy.length === 0 || !rowGrouper) return [undefined, rawRows.length];
+
+    const groupRows = (rows: readonly R[], [groupByKey, ...remainingGroupByKeys]: readonly string[], startRowIndex: number): [GroupByDictionary<R>, number] => {
+      let groupRowsCount = 0;
+      const groups: GroupByDictionary<R> = {};
+      for (const [key, childRows] of Object.entries(rowGrouper(rows, groupByKey))) {
+        // Recursively group each parent group
+        const [childGroups, childRowsCount] = remainingGroupByKeys.length === 0
+          ? [childRows, childRows.length]
+          : groupRows(childRows, remainingGroupByKeys, startRowIndex + groupRowsCount + 1); // 1 for parent row
+        groups[key] = { childRows, childGroups, startRowIndex: startRowIndex + groupRowsCount };
+        groupRowsCount += childRowsCount + 1; // 1 for parent row
+      }
+
+      return [groups, groupRowsCount];
+    };
+
+    return groupRows(rawRows, groupBy, 0);
+  }, [groupBy, rowGrouper, rawRows]);
+
+  const [rows, allGroupRows] = useMemo(() => {
+    const allGroupRows = new Set<unknown>();
+    if (!groupedRows) return [rawRows, allGroupRows];
+
+    const expandGroup = (rows: GroupByDictionary<R> | readonly R[], parentId: string | undefined, level: number): Array<GroupRow<R> | R> => {
+      if (Array.isArray(rows)) return rows;
+      const flattenedRows: Array<R | GroupRow<R>> = [];
```

Comment:
We can move this `flattenedRows` outside `expandGroup`, and just push to it directly.

```patch
diff --git a/src/hooks/useViewportRows.ts b/src/hooks/useViewportRows.ts
index b9769254..b27420e5 100644
--- a/src/hooks/useViewportRows.ts
+++ b/src/hooks/useViewportRows.ts
@@ -47,9 +47,14 @@ export function useViewportRows<R, SR>({
     const allGroupRows = new Set<unknown>();
     if (!groupedRows) return [rawRows, allGroupRows];

-    const expandGroup = (rows: GroupByDictionary<R> | readonly R[], parentId: string | undefined, level: number): Array<GroupRow<R> | R> => {
-      if (Array.isArray(rows)) return rows;
-      const flattenedRows: Array<R | GroupRow<R>> = [];
+    const flattenedRows: Array<R | GroupRow<R>> = [];
+
+    const expandGroup = (rows: GroupByDictionary<R> | readonly R[], parentId: string | undefined, level: number) => {
+      if (Array.isArray(rows)) {
+        flattenedRows.push(...rows);
+        return;
+      }
+
       Object.keys(rows).forEach((groupKey, posInSet, keys) => {
         // TODO: should users have control over the gerenated key?
         const id = parentId !== undefined ? `${parentId}__${groupKey}` : groupKey;
@@ -71,14 +76,13 @@ export function useViewportRows<R, SR>({
         allGroupRows.add(groupRow);

         if (isExpanded) {
-          flattenedRows.push(...expandGroup(childGroups, id, level + 1));
+          expandGroup(childGroups, id, level + 1);
         }
       });
-
-      return flattenedRows;
     };

-    return [expandGroup(groupedRows, undefined, 0), allGroupRows];
+    expandGroup(groupedRows, undefined, 0);
+    return [flattenedRows, allGroupRows];
   }, [expandedGroupIds, groupedRows, rawRows]);

   const isGroupRow = <R>(row: unknown): row is GroupRow<R> => {
```

Reply:
Great idea.